### PR TITLE
Update getdents_syscall tests and comments

### PR DESCRIPTION
**Pull request by rupeshkoushik07**
## Description

This PR adds comments and tests to getdents_syscall
 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- Test A - `lind_project/tests/test_cases/test_a.c`
- Test B - `lind_project/tests/test_cases/test_b.c`

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/tests/fs_tests.rs [hunk 1]
```diff
@@ -1408,6 +1408,92 @@ pub mod fs_tests {
...
+        let mut vec = vec![0u8; bufsize as usize];
+        let baseptr: *mut u8 = &mut vec[0];
+
+        // Create a directory
+        assert_eq!(cage.mkdir_syscall("/getdents", S_IRWXA), 0);
+
+        // Open the directory
+        let fd = cage.open_syscall("/getdents", O_RDWR, S_IRWXA);
+
+        // Attempt to call `getdents_syscall` with an invalid file descriptor
```
Comment (qianxichen233) [inline src/tests/fs_tests.rs:1426]: I think you can test with more types of invalid fds. i.e. non-existing fd, fd larger than MAXFD, etc

File: src/tests/fs_tests.rs [hunk 2]
```diff
@@ -1408,6 +1408,92 @@ pub mod fs_tests {
...
+        // Acquire a lock on TESTMUTEX to prevent other tests from running concurrently,
+        // and also perform clean environment setup.
+        let _thelock = setup::lock_and_init();
+
+        let cage = interface::cagetable_getref(1);
+
+        // Create a regular file
+        let filepath = "/regularfile";
+        let fd = cage.open_syscall(filepath, O_CREAT | O_WRONLY, S_IRWXA);
+        assert_ne!(fd, -(Errno::ENOENT as i32));
```
Comment (qianxichen233) [inline src/tests/fs_tests.rs]: I think changing this to something like `assert!(fd >= 0)` might be better

File: src/tests/fs_tests.rs [hunk 3]
```diff
@@ -1408,6 +1408,92 @@ pub mod fs_tests {
...
+
+        // Allocate a buffer to store directory entries
+        let bufsize = 1024;
+        let mut vec = vec![0u8; bufsize as usize];
+        let baseptr: *mut u8 = &mut vec[0];
+        println!("Buffer contents: {:?}", vec);
```
Comment (qianxichen233) [inline src/tests/fs_tests.rs]: could you remove these print?

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -3496,16 +3496,42 @@ impl Cage {
...
+        // potentially improving performance compared to using the full dirent structure.
         if bufsize <= interface::CLIPPED_DIRENT_SIZE {
             return syscall_error(Errno::EINVAL, "getdents", "Result buffer is too small.");
         }
 
-        let checkedfd = self.get_filedescriptor(fd).unwrap();
+        let checkedfd = match self.get_filedescriptor(fd) {
+            Ok(fd) => fd,
+            Err(_) => return syscall_error(Errno::EBADF, "getdents", "Invalid file descriptor."),
+        };
```
Comment (qianxichen233) [inline src/safeposix/syscalls/fs_calls.rs:3663]: could you add more comments about why some behaviors have to be done for the following code? For example, for `vec_filename.push(DT_UNKNOWN);`, why pushing DT_UNKNOWN is necessary? And there are some similar comments here that only mentioned what it is doing, but did not mention why do this. Would be good if you can complement these comments.

**Review by qianxichen233 — APPROVED** [review]

**Review by yashaswi2000 — APPROVED** [review]
LGTM

File: src/safeposix/syscalls/fs_calls.rs [hunk 5]
```diff
@@ -3650,7 +3690,7 @@ impl Cage {
                                 // convert filename to a filename vector of u8
                                 let mut vec_filename: Vec<u8> = filename.as_bytes().to_vec();
                                 vec_filename.push(b'\0'); // make filename null-terminated
-
+                                // Push DT_UNKNOWN as d_type. This is a placeholder for now, as the actual file type is not yet determined.
```
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs:3693]: some weird whitespace here, have you done a cargo fmt?

File: src/safeposix/syscalls/fs_calls.rs [hunk 6]
```diff
@@ -4548,6 +4590,8 @@ impl Cage {
             // dropping the original reference lets us modify the value without deadlocking
             // the dashmap.
             drop(sementry);
+            // Acquire the semaphore. This operation will block the calling process until the
```
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs]: I'm not sure why this is being changed? I feel like I updated this, is your branch up to date?

**Comment by rennergade** [comment]
Some weird formatting things that I'd like you to take a quick look at but otherwise this is GTG

File: src/safeposix/syscalls/fs_calls.rs (hunk 5 above)
Comment (rupeshkoushik07) [reply src/safeposix/syscalls/fs_calls.rs:3693]: yes even with cargo fmt the whitespace is there

**Review by rennergade — APPROVED** [review]
Approved.

**Pull request merged**